irc/client.go

```diff
@@ -145,58 +147,73 @@ func NewClient(server *Server, conn net.Conn, isTLS bool) {
-		if err != nil {
-			server.logger.Error("internal", "bad server address", err.Error())
-			return
-		}
-		serverPort, _ := strconv.Atoi(serverPortString)
-		clientHost, clientPortString, err := net.SplitHostPort(conn.RemoteAddr().String())
-		if err != nil {
-			server.logger.Error("internal", "bad client address", err.Error())
-			return
+
+	if conn.IsTor {
+		client.SetMode(modes.TLS, true)
+		client.realIP = utils.IPv4LoopbackAddress
+		client.rawHostname = config.Server.TorListeners.Vhost
+	} else {
+		remoteAddr := conn.Conn.RemoteAddr()
+		client.realIP = utils.AddrToIP(remoteAddr)
+		// Set the hostname for this client
+		// (may be overridden by a later PROXY command from stunnel)
+		client.rawHostname = utils.LookupHostname(client.realIP.String())
+		if config.Server.CheckIdent && !utils.AddrIsUnix(remoteAddr) {
+			client.doIdentLookup(conn.Conn)
 		}
-		clientPort, _ := strconv.Atoi(clientPortString)
+	}
 
-		client.Notice(client.t("*** Looking up your username"))
-		resp, err := ident.Query(clientHost, serverPort, clientPort, IdentTimeoutSeconds)
+	client.run()
+}
+
+func (client *Client) doIdentLookup(conn net.Conn) {
+	_, serverPortString, err := net.SplitHostPort(conn.LocalAddr().String())
+	if err != nil {
+		client.server.logger.Error("internal", "bad server address", err.Error())
+		return
+	}
+	serverPort, _ := strconv.Atoi(serverPortString)
+	clientHost, clientPortString, err := net.SplitHostPort(conn.RemoteAddr().String())
+	if err != nil {
+		client.server.logger.Error("internal", "bad client address", err.Error())
+		return
+	}
+	clientPort, _ := strconv.Atoi(clientPortString)
+
+	client.Notice(client.t("*** Looking up your username"))
+	resp, err := ident.Query(clientHost, serverPort, clientPort, IdentTimeoutSeconds)
+	if err == nil {
+		err := client.SetNames(resp.Identifier, "", true)
 		if err == nil {
-			err := client.SetNames(resp.Identifier, "", true)
-			if err == nil {
-				client.Notice(client.t("*** Found your username"))
-				// we don't need to updateNickMask here since nickMask is not used for anything yet
-			} else {
-				client.Notice(client.t("*** Got a malformed username, ignoring"))
-			}
+			client.Notice(client.t("*** Found your username"))
+			// we don't need to updateNickMask here since nickMask is not used for anything yet
 		} else {
-			client.Notice(client.t("*** Could not find your username"))
+			client.Notice(client.t("*** Got a malformed username, ignoring"))
 		}
+	} else {
+		client.Notice(client.t("*** Could not find your username"))
 	}
-	go client.run()
 }
 
 func (client *Client) isAuthorized(config *Config) bool {
 	saslSent := client.account != ""
-	passRequirementMet := (config.Server.passwordBytes == nil) || client.sentPassCommand || (config.Accounts.SkipServerPassword && saslSent)
-	if !passRequirementMet {
+	// PASS requirement
+	if !((config.Server.passwordBytes == nil) || client.sentPassCommand || (config.Accounts.SkipServerPassword && saslSent)) {
```

Comment:
Now this is inline it might be clearer if the conditions are negated instead of the overall thing

```suggestion
	if config.Server.passwordBytes != nil && !client.sentPassCommand && !(config.Accounts.SkipServerPassword && saslSent) {
```